Sources/SwiftFormat/Core/IgnoreFile.swift

```diff
@@ -0,0 +1,93 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+/// A file that describes which files and directories should be ignored by the formatter.
+/// In the future, this file may contain complex rules for ignoring files, based
+/// on pattern matching file paths.
+/// 
+/// Currently, the only valid content for an ignore file is a single asterisk "*",
+/// optionally surrounded by whitespace.
+public class IgnoreFile {
+  /// Name of the ignore file to look for.
+  /// The presence of this file in a directory will cause the formatter
+  /// to skip formatting files in that directory and its subdirectories.
+  fileprivate static let fileName = ".swift-format-ignore"
+
+  /// Errors that can be thrown by the IgnoreFile initializer.
+  public enum Error: Swift.Error {
+    /// Error thrown when an ignore file has invalid content.
+    case invalidContent(URL)
+  }
+
+  /// Create an instance from the contents of the file at the given URL.
+  /// Throws an error if the file content can't be read, or is not valid.
+  public init(contentsOf url: URL) throws {
+    let content = try String(contentsOf: url, encoding: .utf8)
+      guard content.trimmingCharacters(in: .whitespacesAndNewlines) == "*" else {
+        throw Error.invalidContent(url)
+      }
+  } 
+
+  /// Create an instance for the given directory, if a valid
+  /// ignore file with the standard name is found in that directory.
+  /// Returns nil if no ignore file is found.
+  /// Throws an error if an invalid ignore file is found.
+  ///
+  /// Note that this initializer does not search parent directories for ignore files.
+  public convenience init?(forDirectory directory: URL) throws {
+    let url = directory.appendingPathComponent(IgnoreFile.fileName)
+    guard FileManager.default.isReadableFile(atPath: url.path) else {
```

Comment:
Instead of doing two file system accesses to determine if the file is readable and then reading its contents, what do you think about changing `init(contentsOf:)` to take the file contents as a string? Then we could read the file contents here once and pass the contents to the `init(contentsOf:)` initializer (which would need a name matching its new behavior).

It also means that we wouldn’t need to write the ignore files to disk in the tests anymore.

Reply:
Really the only reason for doing two file accesses is to distinguish between the file not existing and other error cases; for clarity it seemed cleaner to do an explicit check. 

We can do that by catching and screening the thrown exception instead if you're that concerned about a double file system access.

Reply:
I've added a String init and eliminated the second access, fwiw.